Data/Semiring/Tropical.hs

```diff
@@ -0,0 +1,253 @@
+#endif
+import Data.Semiring (Semiring(..))
+#if MIN_VERSION_base(4,7,0)
+import Data.Typeable (Typeable)
+#endif
+
+import qualified Data.Monoid as Monoid
+import qualified GHC.Natural as Natural
+
+#if !MIN_VERSION_base(4,6,0)
+-- | On older GHCs, 'Data.Proxy.Proxy' is not polykinded, so we provide our own proxy type for 'Extrema'.
+data EProxy (e :: Extrema) = EProxy
+#endif
+
+-- | A datatype to be used at the kind-level. Its only
+--   purpose is to decide the ordering for the tropical
+--   semiring in a type-safe way.
+data Extrema = Minima | Maxima
+
+-- | The 'Extremum' typeclass exists for us to match on
+--   the kind-level 'Extrema', so that we can recover
+--   which ordering to use in the `Semiring` instance
+--   for 'Tropical`.
+class Extremum (e :: Extrema) where
+  -- unfortunately this has to take a `Proxy` because
+  -- we don't have visual type applications before GHC 8.0
+#if !MIN_VERSION_base(4,6,0)
+  extremum :: EProxy e -> Extrema
+#else
+  extremum :: Proxy e -> Extrema
+#endif
+
+instance Extremum 'Minima where
+  extremum _ = Minima
+  {-# INLINE extremum #-} -- just to be safe
+
+instance Extremum 'Maxima where
+  extremum _ = Maxima
+  {-# INLINE extremum #-} -- just to be safe
+
+-- | The tropical semiring. @'Tropical' ''Minima' a@ is equivalent to the semiring
+--     \[
+--     (a \cup \{+\infty\}, \oplus, \otimes)
+--     \]
+--     where
+--     \[
+--     x \oplus y = min\{x,y\},
+--     \]
+--     \[
+--     x \otimes y = x + y.
+--     \]
+--
+--   @'Tropical' ''Maxima' a@ is equivalent to the semiring
+--     \[
+--     (a \cup \{-\infty\}, \oplus, \otimes)
+--     \]
+--     where
+--     \[
+--     x \oplus y = max\{x,y\},
+--     \]
+--     \[
+--     x \otimes y = x + y.
+--     \]
+--
+-- In literature, the 'Semiring' instance of the 'Tropical' semiring lifts
+-- the underlying semiring's additive structure. One might ask why this lifting doesn't
+-- instead witness isn't 'Monoid', since we only lift 'zero' and 'plus' - the reason is
+-- that usually the additive structure of a semiring is monotonic, i.e.
+-- @a '+' ('min' b c) == 'min' (a '+' b) (a '+' c)@, but in general this is not true.
+-- For example, lifting 'Monoid.Product' 'Natural.Natural' or 'Monoid.Product' 'Word' into 'Tropical' is lawful,
+-- but 'Monoid.Product' 'Int' is not, lacking distributivity: @(-1) '*' ('min' 0 1) '/=' 'min' ((-1) '*' 0) ((-1) '*' 1)@.
+-- So, we deviate from literature and instead
+-- witness the lifting of a 'Monoid', so the user must take care to ensure
+-- that their implementation of 'mappend' is monotonic.
+data Tropical :: Extrema -> * -> * where
+  Infinity :: Tropical e a
+  Tropical :: a -> Tropical e a
+#if MIN_VERSION_base(4,7,0)  
+  deriving (Typeable)
+#endif
+
+#if MIN_VERSION_base(4,7,0)
+deriving instance (Typeable e, Data a) => Data (Tropical e a)
+#endif
+deriving instance Eq a => Eq (Tropical e a)
+deriving instance Foldable (Tropical e)
+deriving instance Functor (Tropical e)
```

Comment:
I am not sure whether treating `Tropical e a` similar to `Maybe a` is a way to go. For instance, it is not entirely expected that `fmap negate Infinity == Infinity`.

Reply:
Ah. I see. Should the instances just not exist? I see no way to write a sane instance (for Functor at least) if this is the case.

Reply:
Yeah, IMHO these instances should not exist.